### PR TITLE
feat(platform): drop support for wildcard capability qualifiers and optional wildcard intention qualifiers

### PR DESCRIPTION
**Pull request by mofogasy**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our [guidelines](https://github.com/SchweizerischeBundesbahnen/scion-microfrontend-platform/blob/master/CONTRIBUTING.md)
- [x] Tests for the changes have been added
- [x] Docs have been added or updated

## PR Type
What kind of change does this PR introduce?

- [ ] Fix
- [x] Feature
- [ ] Documentation
- [ ] Refactoring (changes that neither fixes a bug nor adds a feature)
- [ ] Performance (changes that improve performance)
- [ ] Test (adding missing tests, refactoring tests; no production code change)
- [ ] Chore (other changes like formatting, updating the license, removal of deprecations, etc)
- [ ] Deps (changes related to updating dependencies)
- [ ] CI (changes to our CI configuration files and scripts)
- [ ] Revert (revert of a previous commit)
- [ ] Release (publish a new release)
- [ ] Other... Please describe:

## Issue

Issue Number: #163 

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

Dropping support for wildcard capability qualifiers and optional wildcard intention qualifiers introduced a breaking change in the Intention API.

To migrate, use required parameters instead of asterisk wildcard capability qualifiers and optional parameters instead of optional wildcard capability / intention qualifiers.

### CODE REVIEW

File: projects/scion/microfrontend-platform/src/lib/client/messaging/intent-client.ts [hunk 1]
```diff
@@ -128,20 +128,13 @@ export abstract class IntentClient {
    *  });
    * ```
    *
-   * @param  selector - Allows filtering intents. The qualifier allows using wildcards (such as `*` or `?`) to match multiple intents simultaneously.\
+   * @param  selector - Allows filtering intents. The qualifier allows using the asterisk wildcard character (`*`) to match multiple intents simultaneously.\
    *         Note that the passed filter is only a filter for intents the application is qualified for, i.e., provides a fulfilling capability visible
    *         to the sender.
    *         <p>
    *         <ul>
-   *           <li>**Asterisk wildcard character (`*`):**\
```
Comment (danielwiehl) [inline projects/scion/microfrontend-platform/src/lib/client/messaging/intent-client.ts:136]: I would rephrase the parameter's description, as following:

```code
@param  selector - Allows filtering intents. Note that the passed filter is only a filter for intents the application is qualified for, i.e., provides a fulfilling capability visible to the sender.
        In the qualifier of the filter, you can use the asterisk wildcard character (`*`) to match multiple intents simultaneously.
        - Asterisk wildcard character (*)
          Matches intents with such a qualifier property no matter of its value (except `null` or `undefined`). Use it like this: `{property: '*'}`.
        - Partial wildcard (**)
          Matches intents even if having additional properties. Use it like this: `{'*': '*'}`.
```

File: projects/scion/microfrontend-platform/src/lib/client/messaging/intent-client.ts [hunk 2]
```diff
@@ -185,8 +178,7 @@ export interface IntentSelector {
    */
   type?: string;
   /**
-   * If specified, filters intents matching the given qualifier. You can use the asterisk wildcard (`*`)
-   * or optional wildcard character (`?`) to match multiple intents.
+   * If specified, filters intents matching the given qualifier. You can use the asterisk wildcard (`*`) to match multiple intents.
```
Comment (danielwiehl) [inline projects/scion/microfrontend-platform/src/lib/client/messaging/intent-client.ts:181]: I would be more precise, as following:

```code
If specified, filters intents matching the given qualifier. You can use the asterisk wildcard (`*`) to match multiple intents.
- Asterisk wildcard character (*)
  Matches intents with such a qualifier property no matter of its value (except `null` or `undefined`). Use it like this: `{property: '*'}`.
- Partial wildcard (**)
  Matches intents even if having additional properties. Use it like this: `{'*': '*'}`.
```

File: projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts [hunk 3]
```diff
@@ -72,8 +72,8 @@ export class ɵManifestRegistry implements ManifestRegistry, PreDestroy {
     const filter: ManifestObjectFilter = {appSymbolicName, type: intent.type, qualifier: intent.qualifier || {}};
```
Comment (danielwiehl) [inline projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts:72]: Why are we passing the qualifier in the filter? I am asking because below, we are passing a qualifier predicate.
Or better, I think, here, we do not need to pass a qualifier predicate at all, simply pass the qualifier in the filter.

File: projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts [hunk 4]
```diff
@@ -99,15 +99,23 @@ export class ɵManifestRegistry implements ManifestRegistry, PreDestroy {
    */
   private hasIntentionForCapability(appSymbolicName: string, capability: Capability): boolean {
     const filter: ManifestObjectFilter = {appSymbolicName, type: capability.type, qualifier: capability.qualifier};
```
Comment (danielwiehl) [inline projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts:101]: Same here. Why are we passing the qualifier in the filter?
Or better, I think, here, we do not need to pass a qualifier predicate at all, simply pass the qualifier in the filter.

File: projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts [hunk 5]
```diff
@@ -60,7 +60,7 @@ export class ɵManifestRegistry implements ManifestRegistry, PreDestroy {
   public resolveCapabilitiesByIntent(intent: Intent, appSymbolicName: string): Capability[] {
     assertExactQualifier(intent.qualifier);
     const filter: ManifestObjectFilter = {type: intent.type, qualifier: intent.qualifier || {}};
-    return this._capabilityStore.find(filter, capabilityQualifier => new QualifierMatcher(capabilityQualifier, {evalAsterisk: true, evalOptional: true}).matches(intent.qualifier))
+    return this._capabilityStore.find(filter, capabilityQualifier => new QualifierMatcher(capabilityQualifier, {evalAsterisk: false}).matches(intent.qualifier))
```
Comment (danielwiehl) [inline projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts:63]: Why are we passing the qualifier in the filter? I am asking because below, we are passing a qualifier predicate.
Or better, I think, here, we do not need to pass a qualifier predicate at all, simply pass the qualifier in the filter.

File: projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts [hunk 6]
```diff
@@ -99,15 +99,23 @@ export class ɵManifestRegistry implements ManifestRegistry, PreDestroy {
...
-    if (capability.qualifier && capability.qualifier.hasOwnProperty('*')) {
-      throw Error('[CapabilityRegisterError] Asterisk wildcard (\'*\') not allowed in the qualifier key.');
+    if (capability.qualifier) {
+      if (capability.qualifier.hasOwnProperty('*')) {
+        throw Error('[CapabilityRegisterError] Asterisk wildcard (\'*\') not allowed in the qualifier key.');
+      }
+      if (Object.values(capability.qualifier).some(value => value === '*')) {
+        throw Error('[CapabilityRegisterError] Asterisk wildcard (\'*\') not allowed in the qualifier value. Use required params instead.');
+      }
+      if (Object.values(capability.qualifier).some(value => value === '?')) {
```
Comment (danielwiehl) [inline projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts:116]: Should we remove the check for the optional wildcard in a future release? Then, I suggest creating an issue and putting a TODO here.

File: projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts [hunk 7]
```diff
@@ -137,6 +145,9 @@ export class ɵManifestRegistry implements ManifestRegistry, PreDestroy {
     if (!intention) {
       throw Error(`[IntentionRegisterError] Missing required intention.`);
     }
+    if (intention.qualifier && Object.values(intention.qualifier).some(value => value === '?')) {
```
Comment (danielwiehl) [inline projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts:148]: Same here: Should we remove the check for the optional wildcard in a future release? Then, I suggest creating an issue and putting a TODO here.

File: projects/scion/microfrontend-platform/src/lib/qualifier-matcher.spec.ts [hunk 8]
```diff
@@ -13,63 +13,27 @@ describe('QualifierTester', () => {
...
 
-    it('flags: evalOptional: true, evalAsterisk: true', async () => {
-      const matcher = new QualifierMatcher({entity: 'person', id: '*'}, {evalOptional: true, evalAsterisk: true});
+    it('flags: evalAsterisk: true', async () => {
+      const matcher = new QualifierMatcher({entity: 'person', mode: '*'}, {evalAsterisk: true});
 
-      expect(matcher.matches({entity: 'person', id: '5'})).toBeTrue();
-      expect(matcher.matches({entity: 'person', id: '*'})).toBeTrue();
-      expect(matcher.matches({entity: 'person', id: '?'})).toBeTrue();
+      expect(matcher.matches({entity: 'person', mode: '5'})).toBeTrue();
```
Comment (danielwiehl) [inline projects/scion/microfrontend-platform/src/lib/qualifier-matcher.spec.ts:19]: mode '5'? LOL

File: projects/scion/microfrontend-platform/src/lib/host/manifest-registry/manifest-registry.spec.ts [hunk 9]
```diff
@@ -315,24 +204,24 @@ describe('ManifestRegistry', () => {
...
 
         // Register capabilities of app-2 (public, private, implicit-private)
-        const capabilityId1 = await Beans.get(ManifestRegistry).registerCapability({type: 'view', qualifier: {entity: 'person', id: 'id1'}, private: false}, 'app-2');
-        const capabilityId2 = await Beans.get(ManifestRegistry).registerCapability({type: 'view', qualifier: {entity: 'person', id: 'id2'}, private: true}, 'app-2');
-        const capabilityId3 = await Beans.get(ManifestRegistry).registerCapability({type: 'view', qualifier: {entity: 'person', id: 'id3'}}, 'app-2');
+        const capabilityId1 = await Beans.get(ManifestRegistry).registerCapability({type: 'view', qualifier: {entity: 'person', mode: 'new'}, private: false}, 'app-2');
+        const capabilityId2 = await Beans.get(ManifestRegistry).registerCapability({type: 'view', qualifier: {entity: 'person', mode: 'edit'}, private: true}, 'app-2');
+        const capabilityId3 = await Beans.get(ManifestRegistry).registerCapability({type: 'view', qualifier: {entity: 'person', mode: 'del'}}, 'app-2');
```
Comment (danielwiehl) [inline projects/scion/microfrontend-platform/src/lib/host/manifest-registry/manifest-registry.spec.ts:214]: I suggest replacing `mode: 'del'` with `mode: 'delete'`.

**Review by danielwiehl — CHANGES_REQUESTED** [review]

File: apps/microfrontend-platform-devtools/src/app/dev-tools-manifest.service.ts [hunk 10]
```diff
@@ -104,8 +104,8 @@ export class DevToolsManifestService {
     return this._manifestService.lookupIntentions$({type: capability.type})
       .pipe(
         filterArray(intention => (
-          new QualifierMatcher(intention.qualifier, {evalAsterisk: true, evalOptional: true}).matches(capability.qualifier) ||
-          new QualifierMatcher(capability.qualifier, {evalAsterisk: true, evalOptional: true}).matches(intention.qualifier)
+          new QualifierMatcher(intention.qualifier, {evalAsterisk: true}).matches(capability.qualifier) ||
+          new QualifierMatcher(capability.qualifier, {evalAsterisk: false}).matches(intention.qualifier)
```
[the remark this answers is not in the document]
Comment (mofogasy) [reply apps/microfrontend-platform-devtools/src/app/dev-tools-manifest.service.ts:108]: done

File: apps/microfrontend-platform-devtools/src/app/dev-tools-manifest.service.ts [hunk 11]
```diff
@@ -115,8 +115,8 @@ export class DevToolsManifestService {
     return this._manifestService.lookupCapabilities$({type: intention.type})
       .pipe(
         filterArray(capability => (
-          new QualifierMatcher(intention.qualifier, {evalAsterisk: true, evalOptional: true}).matches(capability.qualifier) ||
-          new QualifierMatcher(capability.qualifier, {evalAsterisk: true, evalOptional: true}).matches(intention.qualifier)
+          new QualifierMatcher(intention.qualifier, {evalAsterisk: true}).matches(capability.qualifier) ||
+          new QualifierMatcher(capability.qualifier, {evalAsterisk: false}).matches(intention.qualifier)
```
[the remark this answers is not in the document]
Comment (mofogasy) [reply apps/microfrontend-platform-devtools/src/app/dev-tools-manifest.service.ts:119]: done

File: projects/scion/microfrontend-platform/src/lib/client/messaging/intent-client.ts (hunk 1 above)
Comment (mofogasy) [reply projects/scion/microfrontend-platform/src/lib/client/messaging/intent-client.ts:136]: done

File: projects/scion/microfrontend-platform/src/lib/client/messaging/intent-client.ts (hunk 2 above)
Comment (mofogasy) [reply projects/scion/microfrontend-platform/src/lib/client/messaging/intent-client.ts:181]: done

File: projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts (hunk 3 above)
Comment (mofogasy) [reply projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts:72]: done

File: projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts (hunk 6 above)
Comment (mofogasy) [reply projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts:116]: done

File: projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts (hunk 7 above)
Comment (mofogasy) [reply projects/scion/microfrontend-platform/src/lib/host/manifest-registry/ɵmanifest-registry.ts:148]: done

File: projects/scion/microfrontend-platform/src/lib/qualifier-matcher.spec.ts (hunk 8 above)
Comment (mofogasy) [reply projects/scion/microfrontend-platform/src/lib/qualifier-matcher.spec.ts:19]: done

File: projects/scion/microfrontend-platform/src/lib/host/manifest-registry/manifest-registry.spec.ts (hunk 9 above)
Comment (mofogasy) [reply projects/scion/microfrontend-platform/src/lib/host/manifest-registry/manifest-registry.spec.ts:214]: done

**Review by danielwiehl — APPROVED** [review]
